### PR TITLE
feat: improve guidance for caching support (#451)

### LINKED ISSUE #451: Document use cases and best practices for cache support.
To support performance in APIs for master data resources, we should describe how to specify caching support.

### PR DESCRIPTION
**Pull request by tkrop**
fixes #451.

The main contributions are in [chapters/performance.adoc:173-309](https://github.com/zalando/restful-api-guidelines/pull/482/files#diff-f409f6bb629f68bac5ace6b270f40a3cR173) and a small extend in [chapters/http-requests.adoc:229-250](https://github.com/zalando/restful-api-guidelines/pull/482/files#diff-c98f51639d3d10ea65dae4c0de86666fR229).

Sorry, that I again mixed some simple reformatting stuff in this commit - especially in chapters/common-headers.adoc. Unfortunately, the diff cannot show this very well.

However, I also mixed an extension to `Prefer` in [chapters/common-headers.adoc:99-155](https://github.com/zalando/restful-api-guidelines/pull/482/files#diff-667f49c57a44c87e7eebac00e5f86cbbR99).

### CODE REVIEW

File: chapters/http-requests.adoc [hunk 1]
```diff
@@ -234,17 +237,17 @@ intended by the operation so that it can be held accountable.
 
 Method implementations must fulfill the following basic properties:
 
-[cols="25%,25%,50%",options="header",]
+[cols="15%,15%,35%,35%",options="header",]
 |====================================================
-| HTTP Method | Safe  | Idempotent
-| {GET}       | {YES} | {YES}
-| {HEAD}      | {YES} | {YES}
-| {POST}      | {NO}  | {AT} No, however you <<229>>.
```
Comment (zeitlinger) [inline chapters/http-requests.adoc]: would rename AT to SHOULD

File: chapters/http-requests.adoc [hunk 2]
```diff
@@ -234,17 +237,17 @@ intended by the operation so that it can be held accountable.
...
-| {PUT}       | {NO}  | {YES}
-| {PATCH}     | {NO}  | {AT} No, however you <<229>>.
-| {DELETE}    | {NO}  | {YES}
-| {OPTIONS}   | {YES} | {YES}
-| {TRACE}     | {YES} | {YES}
+| Method    | Safe  | Idempotent | Cacheable
+| {GET}     | {YES} | {YES}      | {YES}, but <<227>>
+| {HEAD}    | {YES} | {YES}      | {YES}, but <<227>>
+| {POST}    | {NO}  | {AT} No, but <<229>> | {AT} May, but <<227>>
+| {PUT}     | {NO}  | {YES}      | {NO}
```
Comment (zeitlinger) [inline chapters/http-requests.adoc]: why are PUT and DELETE not cacheable?

File: chapters/performance.adoc [hunk 3]
```diff
@@ -4,48 +4,47 @@
...
-resources (see <<182>>
-* pagination for incremental access of larger (result) lists
+* compression of request and response bodies (see <<156>>) 
+* querying field filters to retrieve a subset of resource attributes (see
+  <<157>> below)
+* {ETag} and {If-Match}/{If-None-Match} headers to avoid re-fetching of
+  unchanged resources (see <<182>>)
+* <<pagination>> for incremental access of larger collections of data items
+* caching of master data items, i.e. resources that change rarely or not
+  at all after creation (see <<227>>).
```
Comment (zeitlinger) [inline chapters/performance.adoc]: Is the reference to the Prefer header missing (for minimal or representation)?

File: chapters/http-requests.adoc (hunk 2 above)
Comment (ePaul) [reply chapters/http-requests.adoc]: Because they usually change the state of their resource?
(I guess you could cache a 404 here for some time.)

File: chapters/performance.adoc [hunk 4]
```diff
@@ -4,48 +4,47 @@
...
-Depending on your use case and payload size, you can significantly
-reduce network bandwidth need by supporting filtering of returned entity
-fields. Here, the client can determine the subset of fields he wants to
-receive via the fields query parameter — example see
+Depending on your use case and payload size, you can significantly reduce
+network bandwidth need by supporting filtering of returned entity fields. Here,
+the client can determine the subset of fields he wants to receive via the
+fields query parameter — example see
 https://cloud.google.com/appengine/docs/python/taskqueue/rest/performance#partial-response[Google
 AppEngine API's partial response]:
```
Comment (zeitlinger) [inline chapters/performance.adoc]: it should be "_embed" instead of "_embedded" as we discussed last time (IIRC)

Note that this is not part of this changeset

File: chapters/http-requests.adoc (hunk 2 above)
Comment (zeitlinger) [reply chapters/http-requests.adoc]: yes, 404 for delete and the payload of PUT for subsequent GETs

**Comment by zeitlinger** [comment]
finished the review now - using the diff from IDEA fixed the problems of seeing what has changed.

File: chapters/http-requests.adoc (hunk 1 above)
Comment (tkrop) [reply chapters/http-requests.adoc]: AT is short cut for ATTENTION, which is the icon it refers to.

File: chapters/performance.adoc (hunk 3 above)
Comment (tkrop) [reply chapters/performance.adoc]: Sound like a good extension. I'm inclined to add this. Any objections.

File: chapters/performance.adoc (hunk 4 above)
Comment (tkrop) [reply chapters/performance.adoc]: I have opended #483 to discuss this outside of this PR.

File: chapters/common-headers.adoc [hunk 5]
```diff
@@ -78,66 +69,73 @@ Content-Location: /products/123/images?format=raw
...
 ----
 
-Supporting APIs may return the `Preference-Applied` header also defined
-in https://tools.ietf.org/html/rfc7240[RFC7240] to indicate whether the
-preference was applied.
+Supporting APIs may return the {Preference-Applied} header also defined in
+{RFC7240}[RFC-7240] to indicate whether the preference was applied.
 
 [#182]
 == {MAY} Consider Using `ETag` Together With `If-Match`/`If-None-Match` Header
```
Comment (maxim-tschumak) [inline chapters/common-headers.adoc]: The same argumentation as in the previous comment: let's make it `MAY support ETag`.

Comment (maxim-tschumak) [reply chapters/common-headers.adoc]: I propose to change `use` to `support` in the whole guideline

File: chapters/common-headers.adoc [hunk 6]
```diff
@@ -152,36 +150,39 @@ follows in the API definition:
...
-        The RFC7232 ETag header field in a response provides the current entity-
-        tag for the selected resource. An entity-tag is an opaque identifier for
-        different versions of a resource over time, regardless whether multiple
-        versions are valid at the same time. An entity-tag consists of an opaque
-        quoted string, possibly prefixed by a weakness indicator.
+        The RFC 7232 ETag header field in a response provides the entity-tag of
+        a selected resource. The entity-tag is an opaque identifier for versions
+        and representations of the same resource over time, regardless whether
+        multiple versions are valid at the same time. An entity-tag consists of
+        an opaque quoted string, possibly prefixed by a  weakness indicator (see
```
Comment (maxim-tschumak) [inline chapters/common-headers.adoc]: [very minor] `a  weakness` -> `a weakness` (two whitespaces)

File: chapters/common-headers.adoc [hunk 7]
```diff
@@ -152,36 +150,39 @@ follows in the API definition:
...
+        [RFC 7232 Section 2.3](https://tools.ietf.org/html/rfc7232#section-2.3).
 
       type: string
       required: false
-      example: W/"xy", "5", "7da7a728-f910-11e6-942a-68f728c1ba70"
+      example:  W/"xy", "5", "5db68c06-1a68-11e9-8341-68f728c1ba70"
```
Comment (maxim-tschumak) [inline chapters/common-headers.adoc]: another extra whitespace

File: chapters/http-requests.adoc [hunk 8]
```diff
@@ -226,6 +226,9 @@ An operation can be...
 * [[safe, safe]]{RFC-safe} - the operation semantic is defined to be read-only,
   meaning it must not have _intended side effects_, i.e. changes, to the server
   state.
+* [[cacheable, cacheable]]{RFC-cacheable} - to indicate that responses are
+  allowed to be stored for future reuse. In general, requests to safe methods
+  that do not depend on a current or authoritative response are cacheable.
```
Comment (maxim-tschumak) [inline chapters/http-requests.adoc]: It's not quite clear for me, what is meant under `authoritative response`. Maybe rephrase it a bit?

File: chapters/http-requests.adoc (hunk 2 above)
Comment (maxim-tschumak) [reply chapters/http-requests.adoc]: You would also get a 404 if you are not authorized to delete a resource.

File: chapters/performance.adoc [hunk 9]
```diff
@@ -171,22 +170,140 @@ GET /order/123?embed=(items) HTTP/1.1
...
-transparent HTTP caching should be avoided for RESTful APIs unless the
-API designer has proven to know better.
+As a consequence, client side as well as transparent web caching should be
+avoided, unless the service supports and requires it to protect itself, e.g.
+in case of a heavily used and therefore rate limited master data service, i.e.
+data items that rarely or not at all change after creation.
 
-As default, API providers should set the `Cache-Control: no-cache` header.
+As default, API providers and consumers should always set the {Cache-Control}
+header set to {Cache-Control-no-store} and assume the same setting, if no
```
Comment (maxim-tschumak) [inline chapters/performance.adoc]: should be `Cache-Control: no-store`

File: chapters/performance.adoc [hunk 10]
```diff
@@ -171,22 +170,140 @@ GET /order/123?embed=(items) HTTP/1.1
...
 
-If an API is intended to support caching, it must take care to specify
-this ability by defining the caching boundaries i.e. time-to-live and
-cache constraints, by providing the `Cache-Control` and `Vary` headers
-(Please read https://tools.ietf.org/html/rfc7234[RFC-7234] carefully).
+Caching has to take many aspects into account, e.g. general <<cacheable,
+cacheability>> of response information, our guideline to protect endpoints
+using SSL and <<104, OAuth authorization>>, resource update and invalidation
+rules, existence of multiple consumer instances. As a consequence, caching is
+in best case complex, in worst case inefficient.
```
Comment (maxim-tschumak) [inline chapters/performance.adoc]: I think the biggest peril of using caching is the high probability of becoming **inconsistent** at some point. Should we mention it more explicitly?

File: chapters/performance.adoc [hunk 11]
```diff
@@ -171,22 +170,140 @@ GET /order/123?embed=(items) HTTP/1.1
...
+        a request message, aside the target URL and HTTP method, might have
+        influenced the response. A client or proxy cache must respect this
+        information, to ensure that it delivers the correct cache entry (see
+        [RFC-7231 Section
+        7.1.4](https://tools.ietf.org/html/rfc7231#section-7.1.4)).
+
+      type: string
+      required: false
+      example: "accept-encoding, accept-language"
+  - ETag:
```
Comment (maxim-tschumak) [inline chapters/performance.adoc]: `ETag` header is defined twice. I would prefer to have only one definition (here) and a reference.

File: chapters/performance.adoc [hunk 12]
```diff
@@ -171,22 +170,140 @@ GET /order/123?embed=(items) HTTP/1.1
...
+different media types, or requires other request specific headers. To support
+correct caching you have to carefully choose the value. However, a good first
+default may be:
+
+[source,http]
+----
+Vary: accept, accept-encoding
+----
+
+Anyhow, this is only relevant, if you encourage clients to install generic
```
Comment (maxim-tschumak) [inline chapters/performance.adoc]: It's not relevant for us (Zalando).

File: index.adoc [hunk 13]
```diff
@@ -94,17 +106,40 @@
 
 // Attributes to improve design and linking of common headers.
 :ETag: pass:[<a href="https://tools.ietf.org/html/rfc7232#section-2.3"><code>ETag</code></a>]
+:entity-tag: pass:[<a href="https://tools.ietf.org/html/rfc7232#section-2.3"><code>&lt;entity-tag&gt;</code></a>]
 :If-Match: pass:[<a href="https://tools.ietf.org/html/rfc7232#section-3.1"><code>If-Match</code></a>]
 :If-None-Match: pass:[<a href="https://tools.ietf.org/html/rfc7232#section-3.2"><code>If-None-Match</code></a>]
+
+:Accept: pass:[<a href="https://tools.ietf.org/html/rfc7231#section-5.3.2"><code>Accept</code></a>]
```
Comment (maxim-tschumak) [inline index.adoc]: Does it work for PDFs? We should always look at all produced artifacts.

File: chapters/common-headers.adoc [hunk 14]
```diff
@@ -78,66 +69,73 @@ Content-Location: /products/123/images?format=raw
...
 
-More details in RFC 7231
-https://tools.ietf.org/html/rfc7231#section-7.1.2[7.1.2 Location],
-https://tools.ietf.org/html/rfc7231#section-3.1.4.2[3.1.4.2
-Content-Location]
+More details in RFC 7231 {RFC-7231}#section-7.1.2[7.1.2 Location],
+{RFC-7231}#section-3.1.4.2[3.1.4.2 Content-Location]
 
 [#181]
 == {MAY} Use the Prefer header to indicate processing preferences
```
[the remark this answers is not in the document]
Comment (tkrop) [reply chapters/common-headers.adoc]: Yes good suggestion. I look what I can do.

File: chapters/common-headers.adoc (hunk 5 above)
Comment (tkrop) [reply chapters/common-headers.adoc]: Lets restrict this to the part we are currently working on :)

File: chapters/http-requests.adoc (hunk 8 above)
Comment (tkrop) [reply chapters/http-requests.adoc]: Yes, that is not the easiest sentence: I must admit, I have stolen it from https://tools.ietf.org/html/rfc7231#section-4.2.3 :smile_cat:. What this means is, that you can cache responses for requests to *safe methods*, if you do not need a current answer or an answer from the original server. However, I'm not sure whether this makes the sentence, but I will try some improvement.

File: chapters/performance.adoc (hunk 9 above)
Comment (tkrop) [reply chapters/performance.adoc]: This is source code. I do not like to introduce *attributes* with blanks, but it will be translated as you suggested it.

File: chapters/performance.adoc (hunk 10 above)
Comment (tkrop) [reply chapters/performance.adoc]: In many cases even this is acceptable ...

Do you want to define an action for improvement on this remark?

File: chapters/performance.adoc (hunk 11 above)
Comment (tkrop) [reply chapters/performance.adoc]: Replaced against Link.

File: chapters/performance.adoc (hunk 10 above)
Comment (tkrop) [reply chapters/performance.adoc]: Added *inconsistency* problem more explicit.

File: index.adoc (hunk 13 above)
Comment (tkrop) [reply index.adoc]: Works for me so far. Have we ever checked the general PDF style? We could improve it a lot.

File: chapters/http-requests.adoc [hunk 15]
```diff
@@ -234,17 +238,17 @@ intended by the operation so that it can be held accountable.
 
 Method implementations must fulfill the following basic properties:
```
Comment (tfrauenstein) [inline chapters/http-requests.adoc]: editorial: following basic properties:  --> following basic properties according to https://tools.ietf.org/html/rfc7231[HTTP 1.1 RFC]:

File: chapters/http-requests.adoc (hunk 2 above)
Comment (tfrauenstein) [reply chapters/http-requests.adoc]: fine -- we should not be less restrictive than HTTP 1.1. standard.

File: chapters/http-requests.adoc [hunk 16]
```diff
@@ -234,17 +238,17 @@ intended by the operation so that it can be held accountable.
...
-| {POST}      | {NO}  | {AT} No, however you <<229>>.
-| {PUT}       | {NO}  | {YES}
-| {PATCH}     | {NO}  | {AT} No, however you <<229>>.
-| {DELETE}    | {NO}  | {YES}
-| {OPTIONS}   | {YES} | {YES}
-| {TRACE}     | {YES} | {YES}
+| Method    | Safe  | Idempotent | Cacheable
+| {GET}     | {YES} | {YES}      | {YES}, but <<227>>
+| {HEAD}    | {YES} | {YES}      | {YES}, but <<227>>
+| {POST}    | {NO}  | {AT} No, but <<229>> | {AT} May, but <<227>>
```
Comment (tfrauenstein) [inline chapters/http-requests.adoc]: Yes, "May" according to HTTP standard. 
However, we also should provide following restrictive information here: 

{AT} May, but use cacheable only, if specific POST endpoint is safe (see <<227>>). 
Hint: overwhelming majority of cache implementations only support GET and HEAD.

**Comment by tkrop** [comment]
:+1:

**Comment by tfrauenstein** [comment]
👍

**Comment by regispl** [comment]
Hey, 

The guidance you're giving sounds good to me in general, but there's a couple of things that 'd like to comment on:
1. I'm not sure I get what's the motivation for the detailed "small vs. medium vs. large data size" guidelines - would it make sense to focus on the "best possible" guidance, rather than providing this detailed explanation on case-to-case basis with only slight differences between them (e.g. guidance for small vs. large is almost identical)? So what I mean is putting together whatever is in the "In general (...)" section plus `HEAD` support and pagination (which, by the way, is already a "Must" according to other API Guideline's section) and considering it a "baseline"? 
This is even more confusing when you read the "general" section saying that "you should support ETag Together With If-Match / If-None-Match Header on all cacheable endpoints" but then `ETag`s are - for some reason - specifically mentioned one more time for "small", but omitted for "large". 
In general, while the advices are sound and correct, I find this section a bit too detailed and quite difficult to follow. 
2. Following up on the above - I find the "small vs. medium vs. large" distinction very "subjective" and open for interpretation and as such they don't seem to be useful for the readers. As mentioned above, I'd personally remove it completely (or at least make it way less detailed - one "default" best advice and one or two exceptions listed separately), but if it's to stay, would you consider providing a more detailed guidance on what those mean? 
3. The "Provide efficient methods to warm up and update caches (...)" part makes me think about the "thundering herd" problem on client's startup to me, which always scares me a bit as a service owner. I was wondering if it would make sense to **explicitly** mention the importance of Rate Limiting here (it's mentioned few paragraphs above in slightly different context)? 
4. Related to the point above: would it make sense to "promote" asynchronous eventing here as a possible alternative for using direct REST API calls & cache for accessing datasets (especially the large ones)? While no silver bullet and not the best option for every single case, when it can be used, the benefits of "notify on change" approach are massive - no periodic GETs, no thundering herds on startup to warm-up cache, no problem when cache expires and the service owning the data is down / slow, no cache "management" (respecting different headers, expiry etc.) at all...

Just a few ideas / doubts that came to my mind when reading this section 👍
